docs/book/src/developer/logging.md

```diff
@@ -0,0 +1,158 @@
+correlating lines across logs.
+
+A set of good practices for defining key value pairs is defined in the [Kubernetes Guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#name-arguments), and
+one of the above practices is really important for Cluster API developers
+
+- Developers MUST use `klog.KObj` or `klog.KRef` functions when logging key value pairs for Kubernetes objects, thus
+  ensuring a key value pair representing a Kubernetes object is formatted consistently in all the logs.
+
+Please note that, in order to ensure logs can be easily searched it is important to ensure consistency for the following
+key value pairs (in order of importance):
+
+- Key value pairs identifying the object being reconciled, e.g. a Machine Deployment.
+- Key value pairs identifying the hierarchy of objects being reconciled, e.g. the Cluster a Machine Deployment belongs
+  to.
+- Key value pairs identifying side effects on other objects, e.g. while reconciling a MachineDeployment, the controller 
+  creates a MachinesSet.
+- Other Key value pairs.
+
+## Log Messages
+
+- A Message MUST always start with a capital letter.
+- Period at the end of a message MUST be omitted.
+- Always prefer logging before the action, so in case of errors there will be an immediate, visual correlation between
+  the action log and the corresponding error log; While logging before the action, log verbs should use the -ing form.
+- Ideally log messages should surface a different level of detail according to the target log level (see [log levels](#log-levels)
+  for more details).
+
+## Log Levels
+
+Kubernetes provides a set of [recommendations](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use)
+for log levels; as a small integration on the above guidelines we would like to add:
+
+- Logs at the lower levels of verbosity (<=3) are meant to document “what happened” by describing how an object status
+  is being changed by controller/reconcilers across subsequent reconciliations; as a rule of thumb, it is reasonable
+  to assume that a person reading those logs has a deep knowledge of how the system works, but it should not be required
+  for those persons to have knowledge of the codebase. 
+- Logs at higher levels of verbosity (>=4) are meant to document “how it happened”, providing insight on thorny parts of
+  the code; a person reading those logs usually has deep knowledge of the codebase. 
+- Don’t use verbosity higher than 5.
+
+Ideally, in a future release of Cluster API we will switch to use 2 as a default verbosity for all the Cluster API
+controllers as recommended by the Kubernetes guidelines (currently it is 0).
+
+## Anti patterns
+Developers should not use logs to try to reproduce the same level of details available in an IDE debugger.
```

Comment:
Debatable? The limit of verbosity should be in the `--v` flag level, rather than limiting ourselves from the get-go. For example, a very high level might mean that we can allow tracing in some cases, which is close to an IDE debugger. You can't necessarily attach a debugger in production systems

Reply:
I have dropped this sentence and renamed the paragraph trade-offs
I agree that in some cases it could make sense to get closer to an IDE debugger.